survey/default.nix

```diff
@@ -864,24 +864,55 @@ let
 
   };
 
-
   pkgsWithArchiveFiles = pkgs.extend archiveFilesOverlay;
 
+  fixGhc = ghcPackage0: lib.pipe ghcPackage0 [
+    # musl does not support libdw's alleged need for `dlopen()`, see:
+    #     https://github.com/nh2/static-haskell-nix/pull/116#issuecomment-1585786484
+    #
+    # Nixpkgs has the `enableDwarf` argument only for GHCs versions that are built
+    # with Hadrian (`common-hadrian.nix`), which in nixpkgs is the case for GHC >= 9.6.
+    # So set `enableDwarf = true`, but not for older versions known to not use Hadrian.
+    (ghcPackage:
+      if lib.any (prefix: lib.strings.hasPrefix prefix compiler) ["ghc8" "ghc90" "ghc92" "ghc94"]
+        then ghcPackage # GHC < 9.6, no Hadrian
+        else ghcPackage.override { enableDwarf = false; }
+    )
+    (ghcPackage:
+      ghcPackage.override {
+        enableRelocatedStaticLibs = useArchiveFilesForTemplateHaskell;
+        enableShared = !useArchiveFilesForTemplateHaskell;
+       }
+    )
+  ];
 
-  # This overlay "fixes up" Haskell libraries so that static linking works.
-  # See note "Don't add new packages here" below!
-  haskellLibsReadyForStaticLinkingOverlay = final: previous:
+  setupGhcOverlay = final: previous:
     let
-      previousHaskellPackages =
+      initialHaskellPackages =
         if integer-simple
           # Note we don't have to set the `-finteger-simple` flag for packages that GHC
           # depends on (e.g. text), because nix + GHC already do this for us:
           #   https://github.com/ghc/ghc/blob/ghc-8.4.3-release/ghc.mk#L620-L626
           #   https://github.com/peterhoeg/nixpkgs/commit/50050f3cc9e006daa6800f15a29e258c6e6fa4b3#diff-2f6f8fd152c14d37ebd849aa6382257aR35
-          then previous.haskell.packages.integer-simple."${compiler}"
-          else previous.haskell.packages."${compiler}";
+          then final.haskell.packages.integer-simple."${compiler}"
+          else final.haskell.packages."${compiler}";
     in
       {
+        haskell = previous.haskell // {
```

Comment:
This is approach is based on https://github.com/NixOS/nixpkgs/issues/101580#issuecomment-716086458 — thanks @cdepillabout!